Native E+ CSV output and conditional output file selection

mbadams5 opened the pull request:
Pull request overview
---------------------
This pull request implements a native E+ CSV output capability as well as add a new object to conditionally turn on/off output files.
Fixes #6552 

Overall E+ runtime
| OutPatient Model (seconds) | original/yes | all/yes | original/no | all/no |
|----------------------------|--------------|---------|-------------|--------|
|  9.3 runtime                      | 353.76       | 2941.5  |             |        |
| branch runtime                    | 267.6        | 1662    | 308.4       | 964.8  |
| % reduction                | 24.4%        | 43.5%   | -15.2%      | 41.9%  |
| speed up                   | 1.32x        | 1.77x   | 0.87x       | 1.72x  |

Native E+ vs ReadVarsESO
| OutPatient Model (seconds) | original/yes | all/yes |
|----------------------------|--------------|---------|
|  ReadVarsESO                   | 0.156        | 1782.81 |
| Native E+ CSV                     | .518         | 555.6   |
| % reduction                | -232.1%      | 68.8%   |
| speed up                   | 0.3x         | 3.21x   |

| MediumOffice Model (seconds) | all/yes |
|------------------------------|---------|
| ReadVarsESO                  | 183     |
| Native E+ CSV                | 34.12   |
| % reduction                  | 81.4%   |
| speed up                     | 5.36x   |

* original: OutPatient model as in testfiles
* all: OutPatient model with ALL Output:Variables/Output:Meters from RDD/MDD file
* yes: Outputs all normal output files during simulation
* no: Outputs no output files during simulation using new Output:Control object


### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

mbadams5 committed:
Initial commit of native CSV output

mbadams5 committed:
Update trying to fix monthly output

mbadams5 committed:
Merge remote-tracking branch 'origin/move_bnd_audit_eso_to_fmt' into improve_file_io_with_fmt_refactor

mbadams5 committed:
Add Output:Control object to conditionally turn off output files

mbadams5 committed:
Merge remote-tracking branch 'origin/move_bnd_audit_eso_to_fmt' into improve_file_io_with_fmt_refactor

mbadams5 committed:
Large update which adds most conditional outputs and mtr csv output

mbadams5 committed:
Merge remote-tracking branch 'origin/develop' into improve_file_io_with_fmt_refactor

mbadams5 committed:
Merge remote-tracking branch 'origin/move_bnd_audit_eso_to_fmt' into improve_file_io_with_fmt_refactor

mbadams5 committed:
Fix build error and one integration error

mbadams5 committed:
Fix runtime issues and being adding unit tests

mbadams5 committed:
Fix potential bugs in results framework reporting

mbadams5 committed:
Update to get meteronlyfile working properly

mbadams5 committed:
Merge remote-tracking branch 'origin/develop' into improve_file_io_with_fmt_refactor

mbadams5 commented:
The Mac and Linux CI issues are related to #7907 and not this PR.

mbadams5 committed:
Add YRMeters initialization

mbadams5 committed:
Convert to internal CSV output when ReadVarsESO is selected in CLI

mbadams5 committed:
Fix Windows build issues

mbadams5 committed:
Merge remote-tracking branch 'origin/develop' into improve_file_io_with_fmt_refactor

mbadams5 committed:
Fix issue with ReadWrite files, mainly delight

mbadams5 committed:
Merge remote-tracking branch 'origin/develop' into improve_file_io_with_fmt_refactor

mbadams5 committed:
Merge branch 'address_7907' into improve_file_io_with_fmt_refactor

mbadams5 committed:
Fixing failing integration tests

mbadams5 committed:
Fix bug in Scheduled Output:Variable handling

mbadams5 committed:
Fix a few bugs

mbadams5 committed:
Change name from ResultsSchema to ResultsFramework

mbadams5 committed:
Add CSVWriter class and update ResultsFramework instance name

mbadams5 committed:
Merge remote-tracking branch 'origin/develop' into improve_file_io_with_fmt_refactor

mbadams5 committed:
Fix merge conflicts

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -108149,6 +108149,152 @@ Meter:CustomDecrement,
    A47; \field Output Variable or Meter Name 22
 
+
+Output:Control,
```
@mbadams5 Since this is adding a new object, we should solicit the full team's review of the new inputs.  I don't think we need a separate NFP document, the proposed IDD object speaks for itself. (Sorry I didn't look closely at this sooner.) But we should consider how this fits in with other output objects and @JasonGlazer 's task to add more units flexibility.

In general, I like the goal of this object.  I'd like to make sure we look at all of the related objects to make sure the naming is consistent. And this will need an I/O Ref description and some IDD \notes would be helpful.
`Output:FileControl`, or `OutputControl:Files`may be better for this one. Here's all of the existing objects with Output in the name:
```
Output:DaylightFactors
Output:IlluminanceMap
Output:VariableDictionary
Output:Surfaces:List
Output:Schedules
Output:Constructions
Output:EnergyManagementSystem
Output:Table:SummaryReports
Output:Table:TimeBins
Output:Table:Monthly
Output:Table:Annual
Output:Variable
Output:Meter
Output:Meter:MeterFileOnly
Output:Meter:Cumulative
Output:Meter:Cumulative:MeterFileOnly
Output:JSON
Output:SQLite
Output:EnvironmentalImpactFactors
Output:Diagnostics
Output:DebuggingData
Output:PreprocessorMessage
OutputControl:IlluminanceMap:Style
OutputControl:Sizing:Style
OutputControl:SurfaceColorScheme
OutputControl:Table:Style
OutputControl:ReportingTolerances
```

mbadams5 replied:
@mjwitte I am open to suggestions.
I'll try to add the I/O Ref description and IDD \notes.
I think either of your two name suggestions work. With the way the object currently functions, I'd probably lean towards `OutputControl:Files` but I don't have a strong preference.

JasonGlazer commented:
@mbadams5 have you considered an option to generate the CSV and MTR files in IP units?

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -108149,6 +108149,152 @@ Meter:CustomDecrement,
+       \key Yes
+       \key No
+       \default Yes
+
```
More file types to include: Table, Daylighting Map, VRML (or is that what WRL is above?), SQL, XML, JSON? I'm assuming the goal for this object is to provide one place to enable/disable every possible output file. Maybe that's too ambitious.

mbadams5 replied:
I could definitely add all those files, however, my initial goal was to not add them since they are controlled by other objects. Meaning, if those objects are not present then those outputs are already not output by the simulation. I considered one master enable/disable control object, but I thought that could also lead to user confusion with multiple objects controlling output. However, I am open to suggestions and thoughts.

mjwitte replied:
My thought was that this object could be a way to turn off files, even if they require other objects present to create them. Then one could preserve those settings in the idf but disable the files as desired. If this object is limited to only files that don't have some other object to control/create them, then the list shrinks to eso, audit, zone sizing, system sizing, bnd, mtd, end, and maybe a few others. For example, rdd, mdd, dxf, all have objects that trigger them. 

I'm guessing this object started as a way to activate writing the csv file directly, because there wasn't any place to turn that on/off. So, you could shrink this to OutputControl:TimeSeries, csv yes/no, eso yes/no. But then that overlaps with output:sqlite, and output:json.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -108149,6 +108149,152 @@ Meter:CustomDecrement,
+       \key Yes
+       \key No
+       \default Yes
+  A24, \field Output SCI
```
What's this one?

mjwitte commented:
Also, is the intent here to deprecate ReadVarsESO? And the rvi/mvi functionality to filter and/or specify the order of the csv columns?

mbadams5 commented:
@JasonGlazer I have not considered an option to generate the CSV and MTR files in IP units. I considered that out of scope of this work.

@mjwitte Yes, the intent is to deprecate ReadVarsESO (and in fact it is not even used in this branch anymore on CI since `-r` calls the native CSV now). I forgot to add in rvi/mvi functionality. It shouldn't be too hard (famous last words) to add that functionality with the approach I've taken.

Myoldmopar commented:
There is a use case that this will not directly support -- using one EnergyPlus run's ESO output along with multiple rvi files to generate different report sets.  I'm not saying that it is a dealbreaker or anything, but just note that deprecating ReadVars will end up causing some workflows to require changes.

mbadams5 commented:
@Myoldmopar If i'm understanding that workflow, couldn't this do that if I add in rvi/mvi support? Basically, loop through all rvi/mvi files and generate CSV outputs based on them? Unless I am missing something...

Myoldmopar commented:
How about an object like: `OutputControl:CSV:Configuration` (or something) which takes over responsibility of the RVI/MVI file from the old workflow?  The object, which could have multiple instances in the IDF, declares the output variables to use for this specific CSV, including time averaging, and the name of the output file to write.  It would need to honor the command line output directory option, but this object would then eliminate the need to call out to disk to look for rvi/mvi files which seems like a step backward.  This would be a clean way to ensure these workflows have a clear path forward with what is likely a minimal transition process.

JasonGlazer commented:
@mbadams5 yes I can see that IP unit conversion for the timeseries stuff is outside your scope. Unfortunately, I think the effective speed up for many people will not be realized because of that. Many people are using the workflow

- ESO/MTR (SI) -> convertESOMTR -> ESO/MTR(IP) -> ReadVARS -> CSV (IP) 

and I think that will continue after your direct CSV implementation.

JasonGlazer commented:
And for these users, please don't deprecate ReadVarsESO. It is still needed for people who want IP.

JasonGlazer commented:
Seems like maybe a general overhaul of all these output options for both timeseries and tabular reporting needs to be considered. Perhaps we don't do it all now but we should formulate and work toward a plan that provides the current capabilities plus any additional capabilities that we want and at the same time cleans up the mess of various objects and external utilities.

mbadams5 committed:
Merge remote-tracking branch 'origin/develop' into improve_file_io_with_fmt_refactor

mbadams5 committed:
Fix merge conflicts

# Conflicts:
#	src/EnergyPlus/CommandLineInterface.cc
#	src/EnergyPlus/DataSystemVariables.cc
#	src/EnergyPlus/DaylightingManager.cc
#	src/EnergyPlus/OutputFiles.cc
#	src/EnergyPlus/OutputFiles.hh
#	src/EnergyPlus/OutputProcessor.cc
#	src/EnergyPlus/OutputReportTabular.cc
#	src/EnergyPlus/OutputReports.cc
#	src/EnergyPlus/ScheduleManager.cc
#	src/EnergyPlus/SimulationManager.cc
#	src/EnergyPlus/SizingManager.cc

mbadams5 committed:
Fix build errors

Myoldmopar commented:
@mbadams5 what's your status on this?  I was going to do a code review on it and work to get it in, but before I do, I wanted to make sure you were done wrt the discussions above.  There are conflicts too, but I'm sure most of them are trivial.  If you think the discussion above necessitates a chat, we could do that anytime (next week's iteration call?), or we can just make sure all the threads are resolved here on this PR.  I'm fine either way and I'll look at the code once we're confident it's ready for final review.

mbadams5 committed:
Merge remote-tracking branch 'origin/develop' into improve_file_io_with_fmt_refactor

mbadams5 committed:
Fix merge conflicts and build conflicts

# Conflicts:
#	src/EnergyPlus/CommandLineInterface.cc
#	src/EnergyPlus/DElightManagerF.cc
#	src/EnergyPlus/EMSManager.cc
#	src/EnergyPlus/HeatBalanceManager.cc
#	src/EnergyPlus/OutputFiles.cc
#	src/EnergyPlus/OutputFiles.hh
#	src/EnergyPlus/OutputProcessor.cc
#	src/EnergyPlus/OutputReportTabular.cc
#	src/EnergyPlus/SimulationManager.cc
#	src/EnergyPlus/SizingManager.cc
#	src/EnergyPlus/UtilityRoutines.cc
#	src/EnergyPlus/UtilityRoutines.hh
#	src/EnergyPlus/WindowManager.cc
#	src/EnergyPlus/api/EnergyPlusPgm.cc

mbadams5 committed:
Add tabular,sqlite,json to output control, change object name

mbadams5 committed:
Merge remote-tracking branch 'origin/develop' into improve_file_io_with_fmt_refactor

mbadams5 committed:
Fix merge conflicts

# Conflicts:
#	src/EnergyPlus/OutputFiles.cc
#	src/EnergyPlus/OutputFiles.hh
#	src/EnergyPlus/OutputReportTabular.cc
#	src/EnergyPlus/OutputReportTabular.hh
#	src/EnergyPlus/UtilityRoutines.cc
#	tst/EnergyPlus/unit/OutputReportTabular.unit.cc

mbadams5 committed:
Merge remote-tracking branch 'origin/develop' into improve_file_io_with_fmt_refactor

Myoldmopar commented:
@mbadams5 did you see that something went awry on Windows with that last commit?

mbadams5 committed:
Merge remote-tracking branch 'origin/develop' into improve_file_io_with_fmt_refactor

mbadams5 committed:
Fix merge conflicts

# Conflicts:
#	src/EnergyPlus/DElightManagerF.cc
#	src/EnergyPlus/OutputReportTabular.cc
#	src/EnergyPlus/OutputReportTabular.hh
#	src/EnergyPlus/SolarShading.cc
#	tst/EnergyPlus/unit/OutputReportTabular.unit.cc

mbadams5 committed:
Merge remote-tracking branch 'origin/develop' into improve_file_io_with_fmt_refactor

mbadams5 committed:
Fix merge conflicts

# Conflicts:
#	src/EnergyPlus/CommandLineInterface.cc
#	src/EnergyPlus/DElightManagerF.cc
#	src/EnergyPlus/DataStringGlobals.hh
#	src/EnergyPlus/DataSystemVariables.cc
#	src/EnergyPlus/DaylightingManager.cc
#	src/EnergyPlus/EMSManager.cc
#	src/EnergyPlus/HeatBalanceManager.cc
#	src/EnergyPlus/HeatBalanceSurfaceManager.cc
#	src/EnergyPlus/IOFiles.cc
#	src/EnergyPlus/OutputFiles.hh
#	src/EnergyPlus/OutputProcessor.cc
#	src/EnergyPlus/OutputReportTabular.cc
#	src/EnergyPlus/OutputReports.cc
#	src/EnergyPlus/ResultsFramework.cc
#	src/EnergyPlus/ResultsFramework.hh
#	src/EnergyPlus/ScheduleManager.cc
#	src/EnergyPlus/SimulationManager.cc
#	src/EnergyPlus/SizingManager.cc
#	src/EnergyPlus/StateManagement.cc
#	src/EnergyPlus/UtilityRoutines.cc
#	src/EnergyPlus/WindowManager.cc
#	src/EnergyPlus/api/EnergyPlusPgm.cc
#	tst/EnergyPlus/unit/OutputReportTabular.unit.cc

mbadams5 committed:
Merge remote-tracking branch 'origin/develop' into improve_file_io_with_fmt_refactor

mbadams5 committed:
Revert command line readvars

Myoldmopar commented:
Minor build warning popped up on Mac, but otherwise this last commit is looking **fantastic**.  I have not looked at the changes yet, though.  If we got the build warning fixed up, does this address the discussion during the call yesterday or are there still more changes to make?

mbadams5 committed:
Merge remote-tracking branch 'origin/develop' into improve_file_io_with_fmt_refactor

mbadams5 committed:
Fix build warning

mbadams5 committed:
Fix err file issue

Myoldmopar commented:
Outside of the ERR diffs, this looks good.  I'm assuming we don't actually want those warnings popping up in there, right?  

I think I'm going to make an "IO-Freeze-Release-Candidate" just to make sure packages are doing OK since it has been a while since we made them on CI.  I will plan on closing the IO freeze door by tomorrow morning, so if you can get the ERR diffs cleaned up, that would be great.  If not, we can discuss dropping this in with those temporarily in there.  Assuming it's otherwise ready to go.

Myoldmopar commented on idd/Energy+.idd.in:
```diff
@@ -108530,6 +108530,167 @@ Meter:CustomDecrement,
 
+
+OutputControl:Files,
+	   \memo Conditionally turn on/off output from EnergyPlus.
```
Tab character got left in here.

Myoldmopar commented on src/EnergyPlus/CMakeLists.txt:
```diff
@@ -489,8 +489,8 @@ SET( SRC
-  ResultsSchema.cc
-  ResultsSchema.hh
+  ResultsFramework.cc
+  ResultsFramework.hh
```
:+1:

Myoldmopar reviewed (commented):
This is an incredibly awesome new feature.  I am hopeful that the CI results continue to look promising and we can get this merged in right away.  I think at this point, even if there are a few rough edges, I'm still inclined to get it in and clean up anything remaining with a follow-up issue.

Myoldmopar commented:
Gah, tons of diffs due to develop moving.

mbadams5 committed:
Fix error file reporting

mbadams5 committed:
Merge remote-tracking branch 'origin/develop' into improve_file_io_with_fmt_refactor

mbadams5 committed:
Fix extra tab

Myoldmopar commented:
I don't know where that unit test failure on Mac came from, but I don't think it's the changes made here.  This looks great.  @mbadams5 are you satisfied with where it is at?  If so I'll pull it and test it out a bit.

mbadams5 commented:
@Myoldmopar, yes I think this is ready to go. Please feel free to review it and test it out.

Myoldmopar commented:
OK, so I built and tested this locally.  Some issues, but I think I'm going to merge it in for now to get the I/O side done for freeze.

- I took an example file and tried to run it with the new object and got a seg fault.  It turns out I entered "output:control:files" instead of "outputcontrol:files".  Seg fault because of a bad object name?  
- I changed the object name and it built fine, but the output files were not responding to my inputs.  I found that the get input routine was using the wrong name as well.  In the IDD it is "OutputControl:Files", but in the code it was just "Output:Files".  So it was not actually reading in my new object
- I built with the new name and it was back to a seg fault.  I found that in the input processing function, it was looking for keys in the json input processor with the names "json" and "tabular", instead of "output_json" and "output_tabular".  Seg fault from bad key lookup.
- I then fixed the names it was looking up and it ran, and appeared to generally respond well.  Several of the files were omitted which is great, but the meter file seemed to be produced even if I asked it to not be produced.

Obviously there is still some work to do here, but the I/O is settled, and as long as you don't use the new object, EnergyPlus runs fine.  I'm going to merge and we can open up a follow-up ticket to address these post-freeze.

Thanks @mbadams5, once polished this is going to be a really great addition!

mjwitte commented on src/EnergyPlus/OutputProcessor.cc:
```diff
@@ -8743,34 +8816,20 @@ void ProduceRDDMDD()
-        }
-        mdd_stream << "Program Version," << VerString << ',' << IDDVerString << '\n';
-        mdd_stream << "Var Type (reported time step),Var Report Type,Variable Name [Units]" << '\n';
+        print(state.files.rdd, "Program Version,{},{}{}", VerString, IDDVerString, DataStringGlobals::NL);
+        print(state.files.rdd, "Var Type (reported time step),Var Report Type,Variable Name [Units]{}", DataStringGlobals::NL);
+
+        print(state.files.mdd, "Program Version,{},{}{}", VerString, IDDVerString, DataStringGlobals::NL);
+        print(state.files.mdd, "Var Type (reported time step),Var Report Type,Variable Name [Units]{}", DataStringGlobals::NL);
```
@mbadams5 @Myoldmopar This reintroduced `DataStringGloblas::NL` which was removed in #8178. The rdd and mdd files now have an extra line ending character in the header line `0D 0D 0A`. This causes text editors on Windows to prompt to fix line endings.
